ggml/src/ggml-cuda/common.cuh

```diff
@@ -972,6 +974,133 @@ struct ggml_cuda_graph {
+
+    ggml_cuda_concurrent_event() = default;
+
+    ggml_cuda_concurrent_event(const ggml_cuda_concurrent_event &) = delete;
+    ggml_cuda_concurrent_event & operator=(const ggml_cuda_concurrent_event &) = delete;
+
+    explicit ggml_cuda_concurrent_event(int n_streams) : n_streams(n_streams) {
+        join_events.resize(n_streams);
+
+        for (size_t i = 0; i < join_events.size(); ++i) {
+            CUDA_CHECK(cudaEventCreateWithFlags(&join_events[i], cudaEventDisableTiming));
+        }
+
+        CUDA_CHECK(cudaEventCreateWithFlags(&fork_event, cudaEventDisableTiming));
+    }
+
+    ggml_cuda_concurrent_event(ggml_cuda_concurrent_event && other) noexcept
+    : join_events(std::move(other.join_events))
+    , fork_event(other.fork_event)
+    , n_streams(other.n_streams)
+    , stream_mapping(std::move(other.stream_mapping))
+    , join_node(other.join_node) {
+        other.fork_event = nullptr;
+    }
+
+    // 1. check if all branches don't write to the overlapping memory ranges (except the join node)
+    // 2. check all srcs are either within the branch or outside the event
+    // we assume all nodes have the same buffer
+    bool is_valid() const {
+        std::vector<std::vector<std::pair<int64_t, int64_t>>> write_ranges;
+        write_ranges.resize(n_streams);
+
+        for (const auto & [tensor, stream] : stream_mapping) {
+            const ggml_tensor * t = tensor->view_src ? tensor->view_src : tensor;
+            //concurrent streams begin from 1
+            write_ranges[stream - 1].emplace_back((int64_t) t->data, (int64_t) t->data + ggml_nbytes(t));
+        }
+
+        for (int i = 0; i < n_streams; ++i) {
+            std::sort(write_ranges[i].begin(), write_ranges[i].end());
+        }
+
+        bool writes_overlap = false;
+        bool dependent_srcs = false;
+        for (const auto & [tensor, stream] : stream_mapping) {
+            const ggml_tensor * t = tensor->view_src ? tensor->view_src : tensor;
+
+            // multiple nodes can use join_node's buffer. That is fine because we synchronize on the join node.
+            if (t->data == join_node->data) {
+                continue;
+            }
+            // check if this buffer's write data overlaps with another stream's
+            std::pair<int64_t, int64_t> data_range =
+                std::make_pair((int64_t) t->data, (int64_t) t->data + ggml_nbytes(t));
+            for (int i = 0; i < n_streams; ++i) {
+                if (i == stream - 1) {
+                    continue;
+                }
+                auto it = std::lower_bound(write_ranges[i].begin(), write_ranges[i].end(), data_range);
+
+                if (it != write_ranges[i].end()) {
+                    const std::pair<int64_t, int64_t> & other = *it;
+                    if ((other.first >= data_range.first && other.second <= data_range.second) ||
+                        (other.first <= data_range.first && other.second >= data_range.second)) {
```

Comment:
This logic is not quite correct. Consider the following case as a counterexample: `data_range = {10, 20}, other = {15, 25}`. I would suggest determining which range starts earlier and then checking whether the earlier range ends before the other one starts.